### PR TITLE
Week 18, Project Mongo API - Therese Hagelin

### PR DESCRIPTION
**Pull request by ThereseHag**
I have created a dataset with techsites and created several endpoints for it, some with single result and some returning an array of results. I have also used query params, and tried aggregate to have several queries and sort the results by name. 

Backend is deployed here:
https://techsites.herokuapp.com/

Database is deployed with Mongo DB Atlas Cloud, and a frontend has been deployed as well, as specified in Readme.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
 
 ## The problem
 
-Describe how you approached to problem, and what tools and techniques you used to solve it. How did you plan? What technologies did you use? If you had more time, what would be next?
+This week I started by planning what kind of project I wanted to create, and decided to create a dataset with a collection of techsites with focus on programming. I started with building the dataset, and then started creating RESTful endpoints for the API. I created both a couple of endpoints to display one single result, and several endpoints to display an array of results. I also used query params, to filter on the techsite topic. After doing the more basic queries I also implemented aggregate where I have several query params and sort on name.
+
+After setting up my first endpoints, I created my database with MongoDB and also deployed both my backend and my database. I set up an env file to keep my database user details hidden and made sure to add it to the gitignore file. I then created a corresponding frontend, where I also published my API documentation. 
+
```
Comment (dannuzak) [inline README.md:10]: Good description of your process creating this project! 🦄

File: server.js [hunk 2]
```diff
@@ -1,39 +1,146 @@
...
+  name: String,
+  language: String,
+  type: String,
+  programming_language: String,
+  topic: String,
+  url: String,
+  description: String,
+  offer_training: Boolean,
+  free_or_paid: String,
+});
```
Comment (dannuzak) [inline server.js:26]: Great set of fields!💪🏻

File: server.js [hunk 3]
```diff
@@ -1,39 +1,146 @@
...
+        $sort: {
+          name: 1,
+        },
+      }
+    ]);
+    res.json(techsites);
+  } catch (error) {
+    res.status(400).json({ error: "Something went wrong", details: error })
+  }
+  });
```
Comment (dannuzak) [inline server.js:84]: Great use of regEx, I also learned something here! 🤓

File: server.js [hunk 4]
```diff
@@ -1,39 +1,146 @@
...
+
+  try {
+    const sitesOfLanguage = await Site.find({
+      language: { $regex: "\\b" + language + "\\b", $options: "i" },
+    });
+    res.json(sitesOfLanguage);
+  } catch (error) {
+    res.status(400).json({ error: "Something went wrong", details: error });
+  }
+});
```
Comment (dannuzak) [inline server.js:142]: You experimented and pushed yourself trying several path params, great job! Neat, clean and easy to understand. 🤩

**Review by dannuzak — COMMENTED** [review]
I really enjoyed reading your code Therese! I think you have developed great coding skills and this project is proof of it. It feels very complete as you created your own database, deployed it and it is possible to see it working in the frontend, that's awesome and you should be proud of yourself!

File: README.md [hunk 5]
```diff
@@ -1,13 +1,23 @@
...
+The backend is deployed here: 
+https://techsites.herokuapp.com/
+
+Frontend is deployed here:
+https://techsite-collection.netlify.app/
+
+API documentation can be read here:
+https://techsite-collection.netlify.app/Documentation
+
+And the database is deployed at Mongo DB Cloud Atlas.
```
Comment (KarolCislowski) [inline README.md:23]: short documentation would be nice 🤓

File: server.js [hunk 6]
```diff
@@ -1,39 +1,146 @@
...
+import sites from "./data/tech-sites.json";
+
+dotenv.config();
+
+const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/project-mongo";
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true });
+mongoose.Promise = Promise;
+
+const siteSchema = new mongoose.Schema({
+  siteId: Number,
```
Comment (KarolCislowski) [inline server.js:16]: you will be provided with _id so it is not needed to add a separate one.

File: server.js [hunk 7]
```diff
@@ -1,39 +1,146 @@
...
+            $regex: new RegExp(free_or_paid || "", "i"),
+          },
+        },
+      },
+      {
+        $sort: {
+          name: 1,
+        },
+      }
+    ]);
```
Comment (KarolCislowski) [inline server.js:79]: 💟

File: server.js [hunk 8]
```diff
@@ -1,39 +1,146 @@
...
+  }
+});
+
+//endpoint to display one techsite with the specified name, case insensitive and allow to only use one word in a string.
+app.get("/techsites/name/:name", async (req, res) => {
+  const { name } = req.params;
+
+  try {
+    const singleSite = await Site.findOne({
+      name: { $regex: "\\b" + name + "\\b", $options: "i" },
```
Comment (KarolCislowski) [inline server.js:106]: 👍🏻

**Review by KarolCislowski — APPROVED** [review]
Great API 👍🏻

**Pull request closed without merging**